rfcs/0052-extensible-udt/0052-extensible-udt.md

```diff
@@ -0,0 +1,369 @@
+the owner lock script hash in `args`, this script will be run as an extension
+script. If the script returns success, `is_owner_mode` is set to true. Note, the
+`owner_signature` field can be used by this owner script. Usually, a valid
+signature can be placed here. When tokens are minted, the `owner_script` and
+`owner_signature` can be set to some proper values. When tokens are transferred,
+they can be set to none.
+
+### **xUDT Data**
+
+xUDT data is a molecule serialized `XudtData` structure:
+
+```
+vector Bytes <byte>
+vector BytesVec <Bytes>
+
+table XudtData {
+  lock: Bytes,
+  data: BytesVec,
+}
+```
+
+The `data` field included in `XudtData`, must be of the same length
+as `ScriptVec` structure included in xUDT args. Some extensions might require
+user-specific data stored in each xUDT cell. xUDT data provides a place for such
+data. This XUDT data part can be optional.
+
+The `lock` field included in `XudtData` will not be used by the xUDT script. It
+is reserved for lock script specific data for current cells.
+
+An extension script should first locate the index it resides in xUDT
+args, then look for the data for the current extension script at the same index
+in `data` field of `XudtData` structure.
+
+## **Operations**
+
+xUDT uses the same governance operations as Simple UDT: an owner lock controls
+all governance operations, such as minting.
+
+A normal transfer operation of xUDT, however, differs from Simple UDT. Depending
+on the flags used, there might be 2 usage patterns:
+
+### **Raw Extension Script**
+
+When `flags & 0x1FFFFFFF` are set to 0x1, raw extension data is included in xUDT args directly.
+
+```yaml
+Inputs:
+    <vec> xUDT_Cell
+        Data:
+            <amount: uint128> <xUDT data>
+        Type:
+            code_hash: xUDT type script
+            args: <owner lock script hash> <xUDT args>
+        Lock:
+            <user defined>
+    <...>
+Outputs:
+    <vec> xUDT_Cell
+        Data:
+            <amount: uint128> <xUDT data>
+        Type:
+            code_hash: xUDT type script
+            args: <owner lock script hash> <xUDT args>
+        Lock:
+            <user defined>
+    <...>
+Witnesses:
+    WitnessArgs structure:
+        Lock: <user defined>
+        Input Type: <XudtWitnessInput>
+            owner_script: <None>
+            owner_signature: <None>				
+            raw_extension_data: <None>
+            extension_data: 
+                <vec> BytesVec
+                    <data> 
+                <...>
+```
+
+The witness of the same index as the first input xUDT cell is located by xUDT script. It is parsed first as WitnessArgs structure, the `input_type` or `output_type` field of `WitnessArgs`, is thus treated as `XudtWitnessInput` structure. The length of `extension_data` structure inside must also be the same as `ScriptVec` in `xUDT args`. An extension script might also require transaction-specific data for validation. The witness here provides a place for these data needs.
```

Comment:
I recommend we have a dedicated place explaining the purpose of `extension_data`, instead of hiding it deeply here.